### PR TITLE
workaround for capacity problems of npmjs.org

### PR DESCRIPTION
**Pull request by RubyPaasche**
## Fixes Issue #
this is a workaround for npmjs.org audit-API capacity problems #1685 and #1679.

## Description of Change

Added retry counter for npmjs.org Audit-API requests.

### CODE REVIEW

File: core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java [hunk 1]
```diff
@@ -127,6 +142,14 @@ public NodeAuditSearch(Settings settings) throws MalformedURLException {
                         final NpmAuditParser parser = new NpmAuditParser();
                         return parser.parse(jsonResponse);
                     }
+                case 503:
+                    LOGGER.debug("Invalid payload submitted to Node Audit API. Received response code: {} {}",
+                            conn.getResponseCode(), conn.getResponseMessage());
+                    if (count > 5){
+                        return submitPackage(packageJson, ++count);
```
Comment (jeremylong) [inline core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java:152]: should there be a sleep here?  If we are getting a 503 for capacity issues - shouldn't we be nice and pause a second before retrying?

File: core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java [hunk 2]
```diff
@@ -127,6 +142,14 @@ public NodeAuditSearch(Settings settings) throws MalformedURLException {
                         final NpmAuditParser parser = new NpmAuditParser();
                         return parser.parse(jsonResponse);
                     }
+                case 503:
+                    LOGGER.debug("Invalid payload submitted to Node Audit API. Received response code: {} {}",
```
Comment (jeremylong) [inline core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java]: A 503 is not due to an invalid payload?

```suggestion
                    LOGGER.debug("Node Audit API returned `{} {}` - retrying request.",
```

File: core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java [hunk 3]
```diff
@@ -127,6 +142,14 @@ public NodeAuditSearch(Settings settings) throws MalformedURLException {
...
+                            conn.getResponseCode(), conn.getResponseMessage());
+                    if (count > 5){
+                        return submitPackage(packageJson, ++count);
+                    }
+                    throw new SearchException("Could not perform Node Audit analysis. Invalid payload submitted to Node Audit API.");
```
Comment (jeremylong) [inline core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java]: Again - I don't think we should be calling this an invalid payload?

```suggestion
                    throw new SearchException("Could not perform Node Audit analysis - service returned a 503.");
```

**Review by jeremylong — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java (hunk 2 above)
Comment (RubyPaasche) [reply core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java]: Yeah sorry copy and paste mistake ..

File: core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java (hunk 3 above)
Comment (RubyPaasche) [reply core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java]: Yeah sorry copy and paste mistake ..

File: core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java (hunk 1 above)
Comment (RubyPaasche) [reply core/src/main/java/org/owasp/dependencycheck/data/nodeaudit/NodeAuditSearch.java:152]: I added an "LockSupport.parkNanos" it works the same like Thread.sleep but without Exceptions.
It is an dynamic sleep, it 500 ms multiplied with number of the retry.

**Comment by thib3113** [comment]
Hello @rpaasche , After some research, it seems the node audit results are directly available in json on GitHub .

So, I think it will be really better, to download this vulnerabilities in the local database, and search on it, instead of calling npm audit for each package .

Here you can get it : https://github.com/nodejs/security-advisories
In the ecosystem folder, separate by folder for each dependency, and by vulnerabilities, with semver version vulnerable, and bonus, A cvsscore !!!

Just some more information on this repository, it's a work in progress, because as I read they wants to create a GitHub pages website, to search a package ... But not important in our case .

Other things, this repository, is an extract of this repository : https://github.com/nodejs/security-wg , but better separated ( not all the json in same folder, but separated by package for example ),  and As you can see in this issue : https://github.com/nodejs/security-wg/issues/494 they planned to migrate vulnerabilities to a separate repository. 

So, if you plan to use this, this link can be helpful : 
 - github api to get last master commit : `https://api.github.com/repos/nodejs/security-advisories/commits/master`
 - url to download the zip : `https://github.com/nodejs/security-advisories/archive/master.zip`

I'm not a Java developer, so I can't do a good java code for this (just little things), but if you need some testDatas, or informations, I can help

**Pull request merged**